tsdb/chunks/chunks.go

```diff
@@ -285,17 +295,35 @@ func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
 }
 
 func (w *Writer) WriteChunks(chks ...Meta) error {
-	// Calculate maximum space we need and cut a new segment in case
-	// we don't fit into the current one.
-	maxLen := int64(binary.MaxVarintLen32) // The number of chunks.
-	for _, c := range chks {
-		maxLen += binary.MaxVarintLen32 + 1 // The number of bytes in the chunk and its encoding.
-		maxLen += int64(len(c.Chunk.Bytes()))
-		maxLen += 4 // The 4 bytes of crc32
+	cutChunk := false
+	// First chunk so no need for other checks.
+	if w.wbuf == nil {
+		cutChunk = true
+	} else {
+		// Calculate maximum required space and cut a new segment when
+		// the chunks won't fit in the current segment.
+
+		// Each segment contains: a header + many chunks.
+		// Each chunk contains: data length + encoding + the data itself + crc32
+		reqSpace := int64(SegmentHeaderSize) // The segment header.
```

Comment:
the segment available size is not the full `w.segmentSize`, but always `w.segmentSize - SegmentHeaderSize` , that is why if all chunks total at 100 and the `segmentSize` is 100 they will not fit in this segment because 8 bytes are already used by the segment header.

I think it will be more clear to remove it from here and put it like:
```
if reqSpace > w.segmentSize - SegmentHeaderSize {
```

Let me know if your understanding is different.

Reply:
We are checking `w.n+reqSpace > w.segmentSize`, and the segment size is already included inside `w.n`, so we are double-counting it there. And yes, `if reqSpace > w.segmentSize - SegmentHeaderSize {` would be better.

Reply:
yes but this is different in one case cutting a new segment will help and in the other it won't, but now with your idea to add as many chunks as possible and than cut a new segment there will be no need for this check. very good suggestion and I will refactor soon.

Reply:
after working on this today I realized that we don't have to error even if a single chunk is bigger than the segment size. The segment size is just advisory and it is ok to be bigger than the maximum set size. And it is quite unlikely that a single chunk would be that big to pass the max size of a segment.